### ISSUE TITLE
Specify major, minor, patch version bumps in commit messages

### ISSUE THREAD

**Issue by vinjana** [jgitver/jgitver#39]
Instead of specifying version bumps via tags one could also specify them in commit messages. The advantage would be that one does not even have to keep track of the version numbers. Instead one only says in the commit, whether this is an API-breaking change ("semver: major"), an addition ("semver: minor"), or backwards compatible ("semver: patch"). Such version bumps may then still be added as tags to the git repository.

Note: This would reproduce a feature of [gitversion](https://gitversion.readthedocs.io/) but for Gradle.

**Comment by McFoggy** [comment]
please reopen if needed.

**Comment by michalmela-tomtom** [comment]
Hi @McFoggy! 2 years later, I would like to try and sway you to reconsider adding something similar to the feature mentioned here to `jgitver`.

I have 3 big arguments, if I may:

1. Attain truly automated semantic versioning.

While I understand why you consider the `fix:` / `feat:` prefixes a _pollution_, I believe there is a great merit in basing your semantic versioning on the actual contents of a release (i.e. are we releasing a bunch of fixes? feats? breaking changes?) which is much easier to review and automate than having someone somehow tag the release appropriately (with human errors here much more probable, I believe, than an incorrect commit prefix going past a code review before being merged).

Why have the users run some other tool to derive the new version and an ENV when jgitver could do it and the feature would still be rather cohesive with its current feature set? Speaking of which...

2. All the other cool kids do it.

GitVersion in the C# world does it.
In JS world, there is even a whole specification (and a rather good one I'd say) for the commit patterns: https://www.conventionalcommits.org/en/v1.0.0/ with a related toolset as well: https://github.com/conventional-changelog (I think some Angular-specific toolset does sth similar but his one is more generic).
It looks like Python has some tools following the same standard: https://pypi.org/project/commitizen/ https://pypi.org/project/conventional-commit/

Let's not leave poor old Java behind! I've found some blog posts describing people grafting those tools into Java projects (like having a skeleton `package.json` next to `pom.xml`... my god) but the best that could be said about them "if it looks stupid but it works, then it's not so stupid".

And `jgitver`, I think, is the best candidate to save the day, as it is already the best out of a few similar projects I have found in the JVM ecosystem :)

3. Supporting extreme CD

In the Continuous Deployment world, ain't nobody got time to tag things! Things should just get delivered to production (or as far as your process allows you) as soon as they are committed to the main branch! Yeah! And while we deploy them, it would also be much better to follow the plain old semantic versioning (if we can have our developers follow the commit message patterns) instead of deploying `git describe`-like application versions.

----

Now, this all is just to convince you that the feature makes sense and pick your brain about that. If you green-light the idea, I could try to tackle implementing it.

**Comment by michalmela** [comment]
Hi @McFoggy! 2 years later, I would like to try and sway you to reconsider adding something similar to the feature mentioned here to `jgitver`.

I have 3 big arguments, if I may:

1. Attain truly automated semantic versioning.

While I understand why you consider the `fix:` / `feat:` prefixes a _pollution_, I believe there is a great merit in basing your semantic versioning on the actual contents of a release (i.e. are we releasing a bunch of fixes? feats? breaking changes?) which is much easier to review and automate than having someone somehow tag the release appropriately (with human errors here much more probable, I believe, than an incorrect commit prefix going past a code review before being merged).

Why have the users run some other tool to derive the new version and an ENV when jgitver could do it and the feature would still be rather cohesive with its current feature set? Speaking of which...

2. All the other cool kids do it.

GitVersion in the C# world does it.
In JS world, there is even a whole specification (and a rather good one I'd say) for the commit patterns: https://www.conventionalcommits.org/en/v1.0.0/ with a related toolset as well: https://github.com/conventional-changelog (I think some Angular-specific toolset does sth similar but his one is more generic).
It looks like Python has some tools following the same standard: https://pypi.org/project/commitizen/ https://pypi.org/project/conventional-commit/

Let's not leave poor old Java behind! I've found some blog posts describing people grafting those tools into Java projects (like having a skeleton `package.json` next to `pom.xml`... my god) but the best that could be said about them "if it looks stupid but it works, then it's not so stupid".

And `jgitver`, I think, is the best candidate to save the day, as it is already the best out of a few similar projects I have found in the JVM ecosystem :)

3. Supporting extreme CD

In the Continuous Deployment world, ain't nobody got time to tag things! Things should just get delivered to production (or as far as your process allows you) as soon as they are committed to the main branch! Yeah! And while we deploy them, it would also be much better to follow the plain old semantic versioning (if we can have our developers follow the commit message patterns) instead of deploying `git describe`-like application versions.

----

Now, this all is just to convince you that the feature makes sense and pick your brain about it. If you green-light the idea, I could try to tackle implementing it.
